jerry-core/parser/js/js-parser-statm.c

```diff
@@ -84,6 +73,66 @@ typedef enum
   PARSER_STATEMENT_TRY,
 } parser_statement_type_t;
 
+/**
+ * Parser statement type flags.
+ */
+typedef enum
+{
+  PARSER_STATM_NO_OPTS = 0, /**< no options */
+  PARSER_STATM_SINGLE_STATM = (1 << 0), /**< statment can form single statement context */
+  PARSER_STATM_BREAK_TARGET = (1 << 1), /**< break target statement */
+  PARSER_STATM_CONTINUE_TARGET = (1 << 2), /**< continue target statement */
+  PARSER_STATM_CONTEXT_BREAK = (1 << 3), /**< uses another instruction form when crosses their borders */
+} parser_statement_flags_t;
+
+/**
+ * Parser statement attributes.
+ * Note: the order of the attributes must be keep in sync with parser_statement_type_t
+ */
+static const uint8_t parser_statement_flags[] =
+{
```

Comment:
I like the idea to remove redundancies and simplify the code base. But it would be great if we could make sure that the enum and this array are in synch.

If we can change the order of enum values, we can do it with macros, for example:

```
#define PARSER_STATEMENT_TYPE(F) \
  F(PARSER_STATEMENT_START, PARSER_STATM_NO_OPTS), \
  F(PARSER_STATEMENT_BLOCK, PARSER_STATM_NO_OPTS), \
...

#define PARSER_STATEMENT_TYPE_ES2015(F) ...

#define EXPAND1(A, B)  A
#define EXPAND2(A, B)  B

typedef enum
{
  PARSER_STATEMENT_TYPE(EXPAND1)
#if ENABLED (JERRY_ES2015)
  PARSER_STATEMENT_TYPE_ES2015(EXPAND1)
#endif /* ENABLED (JERRY_ES2015) */
} parser_statement_type_t;

static const uint8_t parser_statement_flags[] =
{
  PARSER_STATEMENT_TYPE(EXPAND2)
#if ENABLED (JERRY_ES2015)
  PARSER_STATEMENT_TYPE_ES2015(EXPAND2)
#endif /* ENABLED (JERRY_ES2015) */
}
```

Reply:
That's a good idea, probably it can be combined with `statement_lengths[]` as well.
@zherczeg What do you think about it?

Reply:
Separate patch. I want to se how it looks like.

Reply:
(lines in the length are quite long, and we have 120 char limit)